### PR TITLE
[Merged by Bors] - Reset payload statuses when resuming fork choice

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

This PR is motivated by a recent consensus failure in Geth where it returned `INVALID` for an `VALID` block. Without this PR, the only way to recover is by re-syncing Lighthouse. Whilst ELs "shouldn't have consensus failures", in reality it's something that we can expect from time to time due to the complex nature of Ethereum. Being able to recover easily will help the network recover and EL devs to troubleshoot.

The risk introduced with this PR is that genuinely INVALID payloads get a "second chance" at being imported. I believe the DoS risk here is negligible since LH needs to be restarted in order to re-process the payload. Furthermore, there's no reason to think that a well-performing EL will accept a truly invalid payload the second-time-around.

## Additional Info

This implementation has the following intricacies:

1. Instead of just resetting *invalid* payloads to optimistic, we'll also reset *valid* payloads. This is an artifact of our existing implementation.
1. We will only reset payload statuses when we detect an invalid payload present in `proto_array`
    - This helps save us from forgetting that all our blocks are valid in the "best case scenario" where there are no invalid blocks.
1. If we fail to revert the payload statuses we'll log a `CRIT` and just continue with a `proto_array` that *does not* have reverted payload statuses.
    - The code to revert statuses needs to deal with balances and proposer-boost, so it's a failure point. This is a defensive measure to avoid introducing new show-stopping bugs to LH.

### CODE REVIEW

File: beacon_node/src/cli.rs [hunk 1]
```diff
@@ -776,4 +776,12 @@ pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
                 .takes_value(true)
                 .default_value("true")
         )
+        .arg(
+            Arg::with_name("reset-payload-statuses")
+                .long("reset-payload-statuses")
+                .help("When present, Lighthouse will forget the payload statuses of any \
+                       already-imported blocks. This can assist in the recovery from a consensus \
+                       failure caused by the execution layer.")
+                .takes_value(true)
```
Comment (michaelsproul) [inline beacon_node/src/cli.rs]:
```suggestion
                .takes_value(false)
```

Just noticed this while lurking

File: consensus/fork_choice/src/fork_choice.rs [hunk 2]
```diff
@@ -1425,15 +1445,56 @@ where
...
+        if matches!(
+            reset_payload_statuses,
+            ResetPayloadStatuses::OnlyWithInvalidPayload
+        ) && !proto_array.contains_invalid_payloads()
+        {
+            return Ok(proto_array);
+        }
+
+        // Reset all blocks back to being "optimistic". This helps recover from an EL consensus
+        // fault where an invalid payload becomes valid.
```
Comment (michaelsproul) [inline consensus/fork_choice/src/fork_choice.rs:1478]: Can we add a debug log here so we know when this is being attempted? I think that might be useful

Comment (paulhauner) [reply consensus/fork_choice/src/fork_choice.rs:1478]: Good call, I made that logger werk in https://github.com/sigp/lighthouse/pull/3498/commits/8551abf258ef37c068d4ce6f1e1d8fcb50e95264

**Review by michaelsproul — APPROVED** [review]
This PR makes minimal changes in the default case while also giving us a nice escape hatch for EL consensus failures. I'm confident that the implementation is correct and won't cause any issues.

Happy to merge

**Comment by paulhauner** [comment]
Penta-batch!

bors r+

**Comment by paulhauner** [comment]
There are conflicts in this batch, but I'll let bors figure it out.

bors r+

**Pull request closed without merging**